--- a/configure
+++ b/configure
@@ -302,6 +302,24 @@
       DISABLE_INCDIR_CHECK=yes
       ;;
 
+   # Exceptional, non-standard case. (ex. luarocks-jit, when LuaJIT must coexist with standard Lua 5.1.)
+   # --------------------------------
+   --config-filename)
+      config_filename="$value"
+      ;;
+
+   --local-dirname)
+      local_dirname="$value"
+      ;;
+
+   --rocks-subdir)
+      rocks_subdir="$value"
+      ;;
+
+   --luarocks-rocks)
+      luarocks_rocks="$value"
+      ;;
+
    # Old options that no longer apply
    # --------------------------------
    --versioned-rocks-dir)
@@ -499,6 +517,10 @@
 LUA_BINDIR=$LUA_BINDIR
 LUA_INCDIR=$LUA_INCDIR
 LUA_LIBDIR=$LUA_LIBDIR
+config_filename=${config_filename:-config-$LUA_VERSION.lua}
+local_dirname=${local_dirname:-.luarocks}
+rocks_subdir=${rocks_subdir:-lib/luarocks/rocks-$LUA_VERSION}
+luarocks_rocks=${luarocks_rocks:-$DATAROOTDIR/lua/$LUA_VERSION}
 FORCE_CONFIG=$FORCE_CONFIG
 EOF
 
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -10,7 +10,10 @@
 
 SHEBANG = \#!$(LUA)
 luarocksconfdir = $(sysconfdir)/luarocks
-luadir = $(datarootdir)/lua/$(LUA_VERSION)
+config_filename ?= config-$(LUA_VERSION).lua
+local_dirname ?= .luarocks
+rocks_subdir ?= lib/luarocks/rocks-$(LUA_VERSION)
+luarocks_rocks ?= $(datarootdir)/lua/$(LUA_VERSION)
 builddir = ./build
 buildbinarydir = ./build-binary
 
@@ -24,19 +27,21 @@
 # Base build
 # ----------------------------------------
 
-build: config.unix $(builddir)/config-$(LUA_VERSION).lua $(builddir)/luarocks $(builddir)/luarocks-admin
+build: config.unix $(builddir)/$(config_filename) $(builddir)/luarocks $(builddir)/luarocks-admin
 
 config.unix:
 	@echo Please run the "./configure" script before building.
 	@echo
 	@exit 1
 
-$(builddir)/config-$(LUA_VERSION).lua: config.unix
+$(builddir)/$(config_filename): config.unix
 	mkdir -p "$(@D)"
 	@printf -- '-- LuaRocks configuration\n\n'\
+	"$$([ "$(local_dirname)" != ".luarocks" ] && printf 'home_tree = home .. "/$(local_dirname)";\\n\\n')"\
+	"$$([ "$(rocks_subdir)" != "lib/luarocks/rocks-$(LUA_VERSION)" ] && printf 'rocks_subdir = "$(rocks_subdir)";\\n\\n')"\
 	'rocks_trees = {\n'\
-	'   { name = "user", root = home .. "/.luarocks" };\n'\
-	"$$([ "$(rocks_tree)" != "$(HOME)/.luarocks" ] && printf '   { name = "system", root = "'"$(rocks_tree)"'" };\\n')"\
+	'   { name = "user", root = home .. "/$(local_dirname)" };\n'\
+	"$$([ "$(rocks_tree)" != "$(HOME)/$(local_dirname)" ] && printf '   { name = "system", root = "'"$(rocks_tree)"'" };\\n')"\
 	'}\n'\
 	'variables = {\n'\
 	"$$([ -n "$(LUA_DIR)" ] && printf '   LUA_DIR = "%s";\\n' "$(LUA_DIR)")"\
@@ -48,9 +53,9 @@
 	'}\n'\
 	> $@
 
-luarocks: config.unix $(builddir)/config-$(LUA_VERSION).lua
-	mkdir -p .luarocks
-	cp $(builddir)/config-$(LUA_VERSION).lua .luarocks/config-$(LUA_VERSION).lua
+luarocks: config.unix $(builddir)/$(config_filename)
+	mkdir -p $(local_dirname)
+	cp $(builddir)/$(config_filename) $(local_dirname)/$(config_filename)
 	rm -f src/luarocks/core/hardcoded.lua
 	echo "#!/bin/sh" > luarocks
 	echo "unset $(LUA_ENV_VARS)" >> luarocks
@@ -70,9 +75,10 @@
 	(printf '$(SHEBANG)\n'\
 	'package.loaded["luarocks.core.hardcoded"] = { '\
 	"$$([ -n "$(FORCE_CONFIG)" ] && printf 'FORCE_CONFIG = true, ')"\
+	'CONFIG_FILENAME = [[$(config_filename)]], '\
 	'SYSCONFDIR = [[$(luarocksconfdir)]] }\n'\
-	'package.path=[[$(luadir)/?.lua;]] .. package.path\n'\
-	'local list = package.searchers or package.loaders; table.insert(list, 1, function(name) if name:match("^luarocks%%.") then return loadfile([[$(luadir)/]] .. name:gsub([[%%.]], [[/]]) .. [[.lua]]) end end)\n'; \
+	'package.path=[[$(luarocks_rocks)/?.lua;]] .. package.path\n'\
+	'local list = package.searchers or package.loaders; table.insert(list, 1, function(name) if name:match("^luarocks%%.") then return loadfile([[$(luarocks_rocks)/]] .. name:gsub([[%%.]], [[/]]) .. [[.lua]]) end end)\n'; \
 	tail -n +2 src/bin/luarocks \
 	)> "$@"
 
@@ -81,9 +87,10 @@
 	(printf '$(SHEBANG)\n'\
 	'package.loaded["luarocks.core.hardcoded"] = { '\
 	"$$([ -n "$(FORCE_CONFIG)" ] && printf 'FORCE_CONFIG = true, ')"\
+	'CONFIG_FILENAME = [[$(config_filename)]], '\
 	'SYSCONFDIR = [[$(luarocksconfdir)]] }\n'\
-	'package.path=[[$(luadir)/?.lua;]] .. package.path\n'\
-	'local list = package.searchers or package.loaders; table.insert(list, 1, function(name) if name:match("^luarocks%%.") then return loadfile([[$(luadir)/]] .. name:gsub([[%%.]], [[/]]) .. [[.lua]]) end end)\n'; \
+	'package.path=[[$(luarocks_rocks)/?.lua;]] .. package.path\n'\
+	'local list = package.searchers or package.loaders; table.insert(list, 1, function(name) if name:match("^luarocks%%.") then return loadfile([[$(luarocks_rocks)/]] .. name:gsub([[%%.]], [[/]]) .. [[.lua]]) end end)\n'; \
 	tail -n +2 src/bin/luarocks-admin \
 	)> "$@"
 
@@ -113,32 +120,32 @@
 	$(INSTALL) '$(builddir)/luarocks-admin' '$(DESTDIR)$(bindir)/luarocks-admin'
 	find src/luarocks/ -type d | while read f; \
 	do \
-	   mkdir -p '$(DESTDIR)$(luadir)'/`echo $$f | sed 's,^src/,,'`; \
+	   mkdir -p '$(DESTDIR)$(luarocks_rocks)'/`echo $$f | sed 's,^src/,,'`; \
 	done
 	find src/luarocks/ -type f -name '*.lua' | while read f; \
 	do \
-	   $(INSTALL_DATA) "$$f" '$(DESTDIR)$(luadir)'/`echo $$f | sed 's,^src/,,'`; \
+	   $(INSTALL_DATA) "$$f" '$(DESTDIR)$(luarocks_rocks)'/`echo $$f | sed 's,^src/,,'`; \
 	done
 ifeq (,$(findstring $(LUA_VERSION),"5.3" "5.4" "5.5"))
 	find src/compat53/ -type d | while read f; \
 	do \
-	   mkdir -p '$(DESTDIR)$(luadir)'/`echo $$f | sed 's,^src/,,'`; \
+	   mkdir -p '$(DESTDIR)$(luarocks_rocks)'/`echo $$f | sed 's,^src/,,'`; \
 	done
 	find src/compat53/ -type f -name '*.lua' | while read f; \
 	do \
-	   $(INSTALL_DATA) "$$f" '$(DESTDIR)$(luadir)'/`echo $$f | sed 's,^src/,,'`; \
+	   $(INSTALL_DATA) "$$f" '$(DESTDIR)$(luarocks_rocks)'/`echo $$f | sed 's,^src/,,'`; \
 	done
 endif
 
 install-config:
 	mkdir -p '$(DESTDIR)$(luarocksconfdir)/'
-	$(INSTALL_DATA)  '$(builddir)/config-$(LUA_VERSION).lua' '$(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua'
+	$(INSTALL_DATA)  '$(builddir)/$(config_filename)' '$(DESTDIR)$(luarocksconfdir)/$(config_filename)'
 
 uninstall:
 	rm -rf $(DESTDIR)$(bindir)/luarocks \
 	       $(DESTDIR)$(bindir)/luarocks-admin \
-	       $(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua \
-	       $(patsubst src/%, $(DESTDIR)$(luadir)/%, $(LUAROCKS_FILES))
+	       $(DESTDIR)$(luarocksconfdir)/$(config_filename) \
+	       $(patsubst src/%, $(DESTDIR)$(luarocks_rocks)/%, $(LUAROCKS_FILES))
 
 # ----------------------------------------
 # Binary install
@@ -148,10 +155,10 @@
 	mkdir -p '$(DESTDIR)$(bindir)/'
 	$(INSTALL) "$(buildbinarydir)/luarocks.exe" "$(DESTDIR)$(bindir)/luarocks"
 	$(INSTALL) "$(buildbinarydir)/luarocks-admin.exe" "$(DESTDIR)$(bindir)/luarocks-admin"
-	mkdir -p '$(DESTDIR)$(luadir)/luarocks/core'
+	mkdir -p '$(DESTDIR)$(luarocks_rocks)/luarocks/core'
 	for f in src/luarocks/core/*.lua src/luarocks/loader.lua; \
 	do \
-	   $(INSTALL_DATA) "$$f" '$(DESTDIR)$(luadir)'/`echo $$f | sed 's,^src/,,'`; \
+	   $(INSTALL_DATA) "$$f" '$(DESTDIR)$(luarocks_rocks)'/`echo $$f | sed 's,^src/,,'`; \
 	done
 
 # ----------------------------------------
--- a/src/luarocks/core/cfg.lua
+++ b/src/luarocks/core/cfg.lua
@@ -725,7 +725,7 @@
    local home_config_file
    local project_config_file
 
-   local config_file_name = "config-"..cfg.lua_version..".lua"
+   local config_file_name = hardcoded.CONFIG_FILENAME or ("config-" .. cfg.lua_version .. ".lua")
 
    do
       local sysconfdir = os.getenv("LUAROCKS_SYSCONFDIR") or hardcoded.SYSCONFDIR
